### PR TITLE
Request identity v1 implementation

### PR DESCRIPTION
**Pull request by jackkleeman**
Implements a simple request signing approach using ed25519 jwts. Signed parameters:
- path as aud (which will include service method)
- issue time, expiry time, and not before time - currently +- 60s

The header looks like:
```
{
  "typ": "JWT",
  "alg": "EdDSA",
  "kid": "publickeyv1_3bQPC1Lsw1PoXFswYssFGUfoDYughmDy5sdY27FobREn"
}
```

The payload looks like:
```
{
  "aud": "/discover",
  "exp": 1712249043,
  "iat": 1712248143,
  "nbf": 1712247243
}
```

The goal of this signing protocol is to prove recent ownership of the private key without requiring mutual TLS. It is not intended to provide strong interception protection, but only client identification; request bodies and headers remain mutable by a MITM attacker. Server-authenticated TLS must still be used to provide interception protection.

A single private key can be given as a local pem file. Public keys in a format like `publickeyv1_CVmG1AvSyedeZpwwd3MRGbRu5yFt3QXXEpQJKyigB9A5` are emitted into logs and can be safely included in service code to aid in verification.

Headers:
1. `x-restate-jwt-v1` -> the jwt
2. `x-restate-signature-scheme` -> `v1`

### CODE REVIEW

**Comment by jackkleeman** [comment]
the unit is seconds, i will clarify and make it a constant - dont see any need to config it yet

will ping runtime folks for a wider review. thx

File: crates/service-client/Cargo.toml [hunk 1]
```diff
@@ -26,16 +26,27 @@ futures = { workspace = true }
...
 serde = { workspace = true }
 serde_json = { workspace = true }
 serde_with = { workspace = true }
 thiserror = { workspace = true }
+tracing = { workspace = true }
+
+pem = "3.0.3"
+ring = "0.17.8"
+bs58 = "0.5.0"
+jsonwebtoken = "9.1.0"
```
Comment (AhmedSoliman) [inline crates/service-client/Cargo.toml:43]: it'd be cool we unify the format to `{ version = "..." }`

File: crates/service-client/src/http.rs [hunk 2]
```diff
@@ -114,14 +132,34 @@ type Connector = ProxyConnector<HttpsConnector<HttpConnector>>;
...
+        let request_identity_key = if let Some(request_identity_private_key_pem_file) =
+            request_identity_private_key_pem_file
+        {
+            Arc::new(ArcSwapOption::from(Some(Arc::new(
+                crate::request_identity::v1::SigningKey::from_pem_file(
+                    request_identity_private_key_pem_file,
+                )?,
+            ))))
+        } else {
+            Arc::new(ArcSwapOption::from(None))
```
Comment (AhmedSoliman) [inline crates/service-client/src/http.rs:153]: nit: `Arc::default()` would probably be enough.

File: crates/service-client/src/http.rs [hunk 3]
```diff
@@ -114,14 +132,34 @@ type Connector = ProxyConnector<HttpsConnector<HttpConnector>>;
...
-    pub fn new(client: hyper::Client<Connector, Body>) -> Self {
-        Self { client }
+    pub fn new(
+        client: hyper::Client<Connector, Body>,
+        request_identity_private_key_pem_file: Option<PathBuf>,
+    ) -> Result<Self, HttpClientError> {
+        let request_identity_key = if let Some(request_identity_private_key_pem_file) =
+            request_identity_private_key_pem_file
+        {
+            Arc::new(ArcSwapOption::from(Some(Arc::new(
```
Comment (AhmedSoliman) [inline crates/service-client/src/http.rs:147]: Curious, Is`ArcSwapOption::new()` not suitable here? or `from_pointee` perhaps?

File: crates/service-client/src/http.rs [hunk 4]
```diff
@@ -204,12 +263,38 @@ impl HttpClient {
...
+}
+
 #[derive(Debug, thiserror::Error)]
 pub enum HttpError {
     #[error(transparent)]
     Hyper(#[from] hyper::Error),
     #[error(transparent)]
     Http(#[from] hyper::http::Error),
+    #[error(transparent)]
+    IdentityV1(#[from] <super::request_identity::v1::Signer<'static> as SignRequest>::Error),
```
Comment (AhmedSoliman) [inline crates/service-client/src/http.rs:297]: neat!

File: crates/service-client/src/request_identity/mod.rs [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+impl<T: SignRequest> SignRequest for Option<T> {
+    type Error = T::Error;
+    fn sign_request(self, mut request: Request<Body>) -> Result<Request<Body>, Self::Error> {
+        match self {
+            Some(signer) => signer.sign_request(request),
+            None => {
+                const SCHEME: HeaderValue = HeaderValue::from_static("unsigned");
+                request.headers_mut().insert(SCHEME_HEADER, SCHEME);
+                Ok(request)
```
Comment (AhmedSoliman) [inline crates/service-client/src/request_identity/mod.rs:22]: Is it necessary to send the extra header in this case?

File: crates/service-client/src/request_identity/v1.rs [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+
+        let keypair = Ed25519KeyPair::from_pkcs8_maybe_unchecked(pem_bytes.as_slice())
+            .map_err(SigningPrivateKeyReadError::KeyRejected)?;
+        let kid = format!(
+            "publickeyv1_{}",
+            bs58::encode(keypair.public_key()).into_string()
+        );
+        let key = jsonwebtoken::EncodingKey::from_ed_der(pem_bytes.as_slice());
+
+        info!(kid, "Loaded request identity key");
```
Comment (AhmedSoliman) [inline crates/service-client/src/request_identity/v1.rs:47]: Bonus if you include the file path.

**Review by AhmedSoliman — APPROVED** [review]
Really nice. I left a couple of tiny optional nits but this is good to go. 

Great job 👏

File: crates/service-client/src/request_identity/mod.rs (hunk 5 above)
Comment (jackkleeman) [reply crates/service-client/src/request_identity/mod.rs:22]: sure, i guess not

File: crates/service-client/src/http.rs (hunk 3 above)
Comment (jackkleeman) [reply crates/service-client/src/http.rs:147]: probably it is yeah :)

**Pull request merged**